contracts/contracts/UserReplicaSetManager.sol

```diff
@@ -0,0 +1,310 @@
+    struct ReplicaSet {
+        uint primary;
+        uint[] secondaries;
+    }
+
+    // Current uint userId to Replica Set
+    mapping (uint => ReplicaSet) userReplicaSets;
+
+    /* Events */
+    event UpdateReplicaSet(
+        uint _userId,
+        uint _primary,
+        uint[] _secondaries
+    );
+
+    event AddOrUpdateCreatorNode(
+        uint _newCnodeId,
+        address _newCnodeDelegateOwnerWallet,
+        uint _proposerSpId
+    );
+
+    event NodeBootstrapAddressUpdated(
+        address _newBootstrapAddress,
+        address _oldBootstrapAddress
+    );
+
+    /* EIP-712 */
+    bytes32 constant ADD_UPDATE_CNODE_REQUEST_TYPEHASH = keccak256(
+        "AddOrUpdateCreatorNode(uint newCnodeId,address newCnodeDelegateOwnerWallet,uint proposerSpId,bytes32 nonce)"
+    );
+    bytes32 constant UPDATE_REPLICA_SET_REQUEST_TYPEHASH = keccak256(
+        "UpdateReplicaSet(uint userId,uint primary,bytes32 secondariesHash,uint oldPrimary,bytes32 oldSecondariesHash,bytes32 nonce)"
+    );
+
+    constructor(
+        address _registryAddress,
+        bytes32 _userFactoryRegistryKey,
+        address _nodeBootstrapAddress,
+        address _userReplicaSetBootstrapAddress,
+        uint _networkId
+    ) SigningLogic("User Replica Set Manager", "1", _networkId) public
+    {
+        require(
+            _registryAddress != address(0x00) &&
+            _userFactoryRegistryKey.length != 0,
+            "Requires non-zero _registryAddress and registryKey"
+        );
+        registry = RegistryInterface(_registryAddress);
+        userFactoryRegistryKey = _userFactoryRegistryKey;
+        nodeBootstrapAddress = _nodeBootstrapAddress;
+        userReplicaSetBootstrapAddress = _userReplicaSetBootstrapAddress;
+    }
+
+    // Chain of trust based authentication scheme
+    // Nodes are required to have an identity in Audius L2 and this function enables
+    //     known entities to register other known entities on L2 contracts.
+    function addOrUpdateCreatorNode(
+        uint _newCnodeId,
+        address _newCnodeDelegateOwnerWallet,
+        uint _proposerSpId,
+        bytes32 _requestNonce,
+        bytes calldata _subjectSig
+    ) external
+    {
+        address signer = _recoverAddOrUpdateCreatorNodeRequestSignerAddress(
+            _newCnodeId,
+            _newCnodeDelegateOwnerWallet,
+            _proposerSpId,
+            _requestNonce,
+            _subjectSig
+        );
+
+        // Requirements for non nodeBootstrapAddress address
+        if (signer != nodeBootstrapAddress) {
+            require(spIdToCreatorNodeDelegateWallet[_proposerSpId] == signer, "Mismatch proposer wallet for existing spID");
+        }
+
+        spIdToCreatorNodeDelegateWallet[_newCnodeId] = _newCnodeDelegateOwnerWallet;
+
+        emit AddOrUpdateCreatorNode(_newCnodeId, _newCnodeDelegateOwnerWallet, _proposerSpId);
+    }
+
+    // TODO: Revisit delete logic - how to remove an spID <-> wallet combo entirely
```

Comment:
TODO: Revisit delete logic - how to remove an spID <-> wallet combo entirely. Is there any actual downside to orphaning old spID <-> wallets?

Reply:
Update here - multi auth now required for any update, so an invalidation can be similarly performed